core/templates/services/exploration-html-formatter.service.ts

```diff
@@ -78,94 +78,113 @@ export class ExplorationHtmlFormatterService {
-      }
-    };
-    if (savedSolution) {
+        element, interactionCustomizationArgs)
+    );
+    // The setAttribute is safe because we verify that the savedSolution
+    // is 'savedMemento()'.
+    if (savedSolution === 'savedMemento()') {
       // TODO(#12292): Refactor this once all interactions have been migrated to
       // Angular 2+, such that we don't need to parse the string in the
       // interaction directives/components.
-      if (spaceToBeAdded) {
-        directiveOuterHtml += ' ';
-      }
+      element.setAttribute('saved-solution', savedSolution);
+      // If interaction is migrated, we only check whether the attribute can be
+      // added, since we cannot add it in proper Angular form using
+      // 'setAttribute'. The rest is done below using string concatenation.
       if (this.migratedInteractions.indexOf(interactionId) >= 0) {
-        directiveOuterHtml += '[saved-solution]="' + savedSolution + '" ';
-      } else {
-        directiveOuterHtml += 'saved-solution="' + savedSolution + '" ';
+        element.removeAttribute('saved-solution');
       }
-      spaceToBeAdded = false;
     }
-    if (labelForFocusTarget) {
-      if (spaceToBeAdded) {
-        directiveOuterHtml += ' ';
-      }
-      directiveOuterHtml += (
-        'label-for-focus-target="' + labelForFocusTarget + '" ');
-      spaceToBeAdded = false;
+    const alphabetRegex = new RegExp('[a-zA-Z]*');
+    // The setAttribute is safe because we verify that the labelForFocusTarget
+    // is only formed of alphabetical characters.
+    if (labelForFocusTarget && alphabetRegex.test(labelForFocusTarget)) {
+      element.setAttribute('label-for-focus-target', labelForFocusTarget);
     }
-    if (spaceToBeAdded) {
-      directiveOuterHtml += ' ';
+    let lastAnswerPropValue = (
+      parentHasLastAnswerProperty ? 'lastAnswer' : 'null'
+    );
+    // The setAttribute is safe because the only possible value is 'lastAnswer'
+    // as per the line above.
+    element.setAttribute('last-answer', lastAnswerPropValue);
+    // If interaction is migrated, we only check whether the attribute can be
+    // added, since we cannot add it in proper Angular form using
+    // 'setAttribute'. The rest is done below using string concatenation.
+    if (this.migratedInteractions.indexOf(interactionId) >= 0) {
+      element.removeAttribute('last-answer');
+    }
+
+    // This is done because 'setAttribute' doesn't allow some characters to be
+    // set as attribute keys (like '[' or ']'). So when interaction is migrated
+    // we first test whether the other parts of the attribute can be added
+    // (code above) and then we add the attribute using string concatenation.
+    if (
+      this.migratedInteractions.indexOf(interactionId) >= 0 &&
+      (lastAnswerPropValue !== null || savedSolution === 'savedMemento()')
+    ) {
+      let interactionHtml = element.outerHTML;
+      const tagEnd = '></oppia-interactive-' + htmlInteractionId + '>';
```

Comment:
Is this stuff permanent or can it be changed in the future to be less "string concatentation"? /cc @srijanreddy98

Reply:
It can be changed with some bigger refactor.

Reply:
Then shouldn't there be a TODO (describing exactly what needs to be done)?

Reply:
For the exactly what needs to be done part I need Srijan.

Reply:
This part doesn't need a TODO. Simply because this method won't work in an Angular2+ scenario. We don't want to compile strings in Angular. That requires shipping the angular compiler which is huge. When are near to migrating this part of the code base, we will create the interactions dynamically. @seanlip might have seen that done in Object Editor component of the schema-based-editor migration PR.

Reply:
But then I still think we need a TODO of some kind to "make sure this is removed", right? I.e. there's clearly further work to be done here, and if it gets sorted out as a side effect of an Angular migration PR, that's great -- but it would be good to have some indication to developers here that this pattern is not one to be followed, and a TODO provides that.

Reply:
/cc @srijanreddy98

Reply:
I have added a TODO referring to an existing issue, but not sure if it is correct.

Reply:
The comment is great and points to the correct issue which will resolve this issue. Thanks Vojta!